From b3e50c3ceb2f583ddb7e504d0ccf221aa7d285a8 Mon Sep 17 00:00:00 2001
From: limeng0715 <limeng0715@thundersoft.com>
Date: Mon, 24 Nov 2025 14:00:47 +0800
Subject: [PATCH] audio: pipewire: modify pipewire configuration for rubikpi3

1. Modify pa-pal-plugins to support HDMI out audio.
2. Based on the Rubikpi3 hardware design, the pipewire-plugin
has been modified to support HDMI out and headset playback,
and also supports headset mic recording.
3. Pipewire default sink interface is HDMI out.

Signed-off-by: limeng0715 <limeng0715@thundersoft.com>
---
 modules/pa-pal-plugins/module-pal-card/configs/qcm6490-idp-snd-card.conf         | 22 ++++++++++++++-----
 modules/pa-pal-plugins/module-pal-card/src.c           |  2 +-
 pipewire-plugin/pw-pal-plugin.c               |  6 +++--
 pipewire-plugin/pw-pal-plugin.conf            | 18 ++++-----------
 4 files changed, 26 insertions(+), 22 deletions(-)

diff --git a/modules/pa-pal-plugins/module-pal-card/configs/qcm6490-idp-snd-card.conf b/modules/pa-pal-plugins/module-pal-card/configs/qcm6490-idp-snd-card.conf
index 182d1dc..4dc004a 100644
--- a/modules/pa-pal-plugins/module-pal-card/configs/qcm6490-idp-snd-card.conf
+++ b/modules/pa-pal-plugins/module-pal-card/configs/qcm6490-idp-snd-card.conf
@@ -1,6 +1,18 @@
 [Global]
 default-profile = default
 
+[Port hdmi]
+description = hdmi
+default-channel-map = front-left,front-right
+default-sample-rate = 48000
+direction = out
+priority = 150
+presence = always
+device = PAL_DEVICE_OUT_HDMI
+encodings = pcm
+sample-formats = s32le
+channel-maps = front-left,front-right
+
 [Port speaker]
 description = speaker
 default-channel-map = front-left,front-right
@@ -94,7 +106,7 @@ description = Default pal profile
 priority = 500
 max-sink-channels = 2
 max-source-channels = 2
-port-names = speaker speaker-mic handset-mic headset headset-mic bta2dp-out btsco-in btsco-out
+port-names = speaker speaker-mic handset-mic headset headset-mic bta2dp-out btsco-in btsco-out hdmi
 
 [Sink low-latency0]
 description = pal sink to play via low-latency path
@@ -106,7 +118,7 @@ default-channel-map = front-left,front-right
 default-buffer-size = 1024
 default-buffer-count = 4
 encodings = pcm
-port-names = speaker headset
+port-names = speaker headset hdmi
 presence = always
 avoid-processing = rate channels
 use-hw-volume = true
@@ -122,7 +134,7 @@ default-buffer-size = 1920
 default-buffer-count = 4
 encodings = pcm
 avoid-processing = rate channels
-port-names = speaker headset bta2dp-out
+port-names = speaker headset bta2dp-out hdmi
 presence = always
 use-hw-volume = true
 
@@ -132,7 +144,7 @@ type = PAL_STREAM_COMPRESSED
 default-encoding = pcm
 default-sample-rate = 48000
 default-sample-format = s16le
-default-channel-map = front-left,front-right
+default-channel-map = front-left,front-right hdmi
 encodings = mpeg aac
 default-buffer-size = 16484
 default-buffer-count = 4
@@ -149,7 +161,7 @@ default-sample-format = s16le
 default-channel-map = front-left,front-right
 default-buffer-size = 1024
 default-buffer-count = 4
-port-names = speaker headset btsco-out
+port-names = speaker headset btsco-out hdmi
 presence = always
 use-hw-volume = true
 
diff --git a/modules/pa-pal-plugins/module-pal-card/src/pal-utils.c b/modules/pa-pal-plugins/module-pal-card/src/pal-utils.c
index e0642eb..5318ae2 100644
--- a/modules/pa-pal-plugins/module-pal-card/src/pal-utils.c
+++ b/modules/pa-pal-plugins/module-pal-card/src/pal-utils.c
@@ -57,7 +57,7 @@ pa_pal_util_port_to_pal_device_mapping port_to_pal_device[] = {
     { (char *)"speaker-mic",      PAL_DEVICE_IN_SPEAKER_MIC,            (char *)"PAL_DEVICE_IN_SPEAKER_MIC" },
     { (char *)"headset-mic",      PAL_DEVICE_IN_WIRED_HEADSET,          (char *)"PAL_DEVICE_IN_WIRED_HEADSET" },
     { (char *)"linein",           PAL_DEVICE_IN_LINE,                   (char *)"PAL_DEVICE_IN_LINE" },
-    { (char *)"hdmi-out",         PAL_DEVICE_OUT_AUX_DIGITAL,           (char *)"PAL_DEVICE_OUT_AUX_DIGITAL" },
+    { (char *)"hdmi-out",         PAL_DEVICE_OUT_HDMI,                  (char *)"PAL_DEVICE_OUT_HDMI" },
     { (char *)"bta2dp-in",        PAL_DEVICE_IN_BLUETOOTH_A2DP,         (char *)"PAL_DEVICE_IN_BLUETOOTH_A2DP" },
     { (char *)"btsco-in",         PAL_DEVICE_IN_BLUETOOTH_SCO_HEADSET,  (char *)"PAL_DEVICE_IN_BLUETOOTH_SCO_HEADSET" },
     { (char *)"btsco-out",        PAL_DEVICE_OUT_BLUETOOTH_SCO,         (char *)"PAL_DEVICE_OUT_BLUETOOTH_SCO" },
diff --git a/pipewire-plugin/pw-pal-plugin.c b/pipewire-plugin/pw-pal-plugin.c
index c9f2818..d471c81 100644
--- a/pipewire-plugin/pw-pal-plugin.c
+++ b/pipewire-plugin/pw-pal-plugin.c
@@ -596,16 +596,18 @@ int pipewire__module_init(struct pw_impl_module *module, const char *args)
     if (value) {
         if (strstr(value, "Sink")) {
             udata->isplayback = true;
-            udata->pal_device_id = PAL_DEVICE_OUT_SPEAKER;
+            udata->pal_device_id = PAL_DEVICE_OUT_HDMI;
         }
         else {
             udata->isplayback = false;
-            udata->pal_device_id = PAL_DEVICE_IN_SPEAKER_MIC;
+            udata->pal_device_id = PAL_DEVICE_IN_WIRED_HEADSET;
         }
     }
 
     value = pw_properties_get(props, PW_KEY_NODE_NAME);
     if (value) {
+        if (strstr(value, "pal_sink_hdmi"))
+            udata->pal_device_id = PAL_DEVICE_OUT_HDMI;
         if (strstr(value, "pal_sink_speaker"))
             udata->pal_device_id = PAL_DEVICE_OUT_SPEAKER;
         else if (strstr(value, "pal_sink_headset"))
diff --git a/pipewire-plugin/pw-pal-plugin.conf b/pipewire-plugin/pw-pal-plugin.conf
index f5712cd..4a425fd 100644
--- a/pipewire-plugin/pw-pal-plugin.conf
+++ b/pipewire-plugin/pw-pal-plugin.conf
@@ -1,8 +1,8 @@
 context.modules = [
 {   name = libpipewire-module-pal
     args = {
-        node.name = "pal_sink_speaker_ll"
-        node.description = "pal sink speaker ll"
+        node.name = "pal_sink_hdmi_ll"
+        node.description = "pal sink hdmi_ll"
         stream.props = {
             audio.position = [ FL FR ]
         }
@@ -12,8 +12,8 @@ context.modules = [
 },
 {   name = libpipewire-module-pal
     args = {
-        node.name = "pal_sink_speaker_db"
-        node.description = "pal sink speaker_db"
+        node.name = "pal_sink_hdmi_db"
+        node.description = "pal sink hdmi_db"
         stream.props = {
             audio.position = [ FL FR ]
         }
@@ -43,16 +43,6 @@ context.modules = [
         media.role = "music"
     }
 },
-{   name = libpipewire-module-pal
-    args = {
-        node.name = "pal_source_speaker_mic"
-        node.description = "pal source handset mic"
-        stream.props = {
-            audio.position = [ FL FR ]
-        }
-        media.class = "Audio/Source"
-    }
-},
 {   name = libpipewire-module-pal
     args = {
         node.name = "pal_source_headset_mic"
-- 
2.17.1

